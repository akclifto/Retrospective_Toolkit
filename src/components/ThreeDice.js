import React, { Suspense } from 'react'
import { TextureLoader } from 'three'
import { Canvas, useLoader } from 'react-three-fiber'
import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader'
import { Html, draco, OrbitControls } from 'drei'
import { Physics, useBox } from 'use-cannon'
import { themes } from '../constants/DieConstants'

// work in progress
const ThemedDie = (props) => {
  if (props.theme === 'action') {
    return (
      <Die position={props.position} images={themes.action.images} />
    )
  }
}

const Die = (props) => {
  const [mesh, api] = useBox(() => ({ mass: 1, position: props.position, velocity: [5, 0, -3], angularVelocity: [5, 0, 3] }))
  return (
    <mesh ref={mesh}>
      <boxBufferGeometry attach='geometry' args={[1, 1, 1]} />
      {props.images.map((image, key) => (
        // eslint-disable-next-line react-hooks/rules-of-hooks
        <meshStandardMaterial key={key} attachArray='material' map={useLoader(TextureLoader, image)} />
      ))}
      {/* <Theme theme={props.theme} /> */}
    </mesh>
  )
}

const Plane = () => {
  return (
    <mesh>
      <planeBufferGeometry attach='geometry' args={[100, 100]} />
      <meshStandardMaterial attach='material' color='red' />
    </mesh>
  )
}

function Model ({ url, props }) {
  const [tray] = useBox(() => ({
    type: 'Static',
    args: [12, 0, 22],
    position: [0, 1, 0],
    rotation: [0, -Math.PI / 2, 0],
    ...props
  }))
  const { scene } = useLoader(GLTFLoader, url, draco())
  return (
    <primitive ref={tray} object={scene} dispose={null} />
  )
}

const Lights = () => {
  return (
    <>
      <ambientLight intensity={0.5} />
      <directionalLight
        intensity={1.0}
        position={[0, 20, 0]}
      />
    </>
  )
}

const ThreeDice = (props) => {
  return (
    <Canvas concurrent style={{ width: '100vw', height: '500px' }} camera={{ position: [0, 20, 12], fov: 50 }}>
      <Physics>
        <Suspense fallback={<Html>loading..</Html>}>
          <ThemedDie position={[-6, 10, 3]} theme='action' />
          <ThemedDie position={[-5, 9, 4]} theme='action' />
          <ThemedDie position={[-7, 10, 3]} theme='action' />
          <ThemedDie position={[-5, 10, 3]} theme='action' />
          <ThemedDie position={[-6, 15, 3]} theme='action' />
          <Model url='trayModel/tray.glb' />
        </Suspense>
      </Physics>
      <Lights />
      <gridHelper args={[50, 50, 'red', 'black']} />
      <OrbitControls />
    </Canvas>
  )
}

export default ThreeDice
