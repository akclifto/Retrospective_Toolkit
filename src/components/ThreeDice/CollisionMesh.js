/* istanbul ignore file */
import React from "react";
import { useBox } from "@react-three/cannon";

const CollisionMesh = () => {
  const [floor] = useBox(() => ({
    position: [0.22, 0, 0.3],
    rotation: [0, -Math.PI / 2, 0],
    args: [18.5, 2.43, 36.5],
    material: { friction: 1000 },
  }));
  const [left] = useBox(() => ({
    position: [-18.05, 2, 0.35],
    rotation: [0, -Math.PI / 2, 0],
    args: [20, 4, 1],
  }));
  const [right] = useBox(() => ({
    position: [18.48, 2, 0.35],
    rotation: [0, -Math.PI / 2, 0],
    args: [16, 4, 0],
  }));
  const [bottom] = useBox(() => ({
    position: [0.22, 2, 9.53],
    args: [37, 4, 0],
  }));
  const [top] = useBox(() => ({
    position: [0.22, 2, -8.73],
    args: [37, 4, 0],
  }));
  return (
    <group>
      <mesh ref={floor}>
        <boxBufferGeometry args={[18.5, 0, 36.5]} />
        <meshStandardMaterial color="yellow" transparent opacity={0} />
      </mesh>
      <mesh ref={left}>
        <boxBufferGeometry args={[18.36, 4, 0]} />
        <meshStandardMaterial color="red" transparent opacity={0} />
      </mesh>
      <mesh ref={right}>
        <boxBufferGeometry args={[18.36, 4, 0]} />
        <meshStandardMaterial color="red" transparent opacity={0} />
      </mesh>
      <mesh ref={bottom}>
        <boxBufferGeometry args={[36.52, 4, 0]} />
        <meshStandardMaterial color="green" transparent opacity={0} />
      </mesh>
      <mesh ref={top}>
        <boxBufferGeometry args={[36.52, 4, 0]} />
        <meshStandardMaterial color="green" transparent opacity={0} />
      </mesh>
    </group>
  );
};

export default CollisionMesh;
