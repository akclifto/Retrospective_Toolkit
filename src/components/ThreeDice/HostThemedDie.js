import React, { useEffect } from "react";
import { useTexture } from "@react-three/drei";
import { useBox } from "@react-three/cannon";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import { useGesture, useDrag } from "react-use-gesture";

/* istanbul ignore next */
const singleRollSound = () => {
  new Audio("/diceRoll.m4a").play();
};

/* istanbul ignore next */
const ThemedDie = (props) => {
  const {
    diceInitPos,
    rerollValue,
    rerollDieToggle,
    imageSet,
    geom,
    setOrbitControl,
    mousePos,
    rotationValues,
  } = props;
  const textures = useTexture([...imageSet]);

  const [mesh, api] = useBox(() => ({
    mass: 300,
    inertia: 13,
    position: [diceInitPos[0], diceInitPos[1], diceInitPos[2]],
    rotation: [
      rotationValues[0] * Math.PI,
      rotationValues[1] * Math.PI,
      rotationValues[2] * Math.PI,
    ],
    linearDamping: 0.5,
    angularDamping: 0.1,
    material: { restitution: 0.3 },
  }));

  const isDragging = React.useRef(false);

  const clickBind = useGesture({
    onClick: () => {
      if (!isDragging.current) {
        rerollDieToggle(!rerollValue);
        singleRollSound();
      }
    },
  });

  const dragBind = useDrag(
    ({ first, last }) => {
      if (first) {
        setOrbitControl(false);
        isDragging.current = true;
        api.mass.set(0);
        api.collisionResponse.set(0);
      } else if (last) {
        // eslint-disable-next-line no-return-assign
        setTimeout(() => (isDragging.current = false), 100);
        setOrbitControl(true);
        api.mass.set(300);
        api.collisionResponse.set(1);
      }
      api.position.set(mousePos[0], 2, -mousePos[1]);
    },
    {
      initial: () => [mousePos[0], mousePos[1]],
      filterTaps: true,
    }
  );

  useEffect(() => {
    api.position.set(diceInitPos[0], diceInitPos[1], diceInitPos[2]);
    api.velocity.set(15, 0, -10);
    api.angularVelocity.set(-15, 2, -10);
    api.rotation.set(
      rotationValues[0] * Math.PI,
      rotationValues[1] * Math.PI,
      rotationValues[2] * Math.PI
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rotationValues]);

  return (
    <mesh
      ref={mesh}
      geometry={geom}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dragBind()}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...clickBind()}
    >
      <boxBufferGeometry />
      {textures.map((image) => (
        <meshStandardMaterial
          key={uuidv4()}
          flatShading
          roughness={0.8}
          attachArray="material"
          map={image}
        />
      ))}
    </mesh>
  );
};

ThemedDie.propTypes = {
  diceInitPos: PropTypes.arrayOf(PropTypes.number).isRequired,
  imageSet: PropTypes.arrayOf(PropTypes.string).isRequired,
  rerollValue: PropTypes.bool.isRequired,
  rerollDieToggle: PropTypes.func.isRequired,
  geom: PropTypes.shape().isRequired,
  setOrbitControl: PropTypes.func.isRequired,
  mousePos: PropTypes.arrayOf(PropTypes.number).isRequired,
  rotationValues: PropTypes.arrayOf(PropTypes.number).isRequired,
};

export default ThemedDie;
