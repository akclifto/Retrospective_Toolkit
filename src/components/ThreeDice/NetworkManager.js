/* eslint-disable no-console */
import { useState, useEffect } from "react";
import { v4 as uuidv4 } from "uuid";
import * as io from "socket.io-client";

const generateRoll = () => {
  // some logic here
};

const sendRoll = () => {
  // some logic here
};

const getRoll = () => {
  // some logic here
};

const NetworkManager = () => {
  const [socket, setSocket] = useState();
  let roomId;
  console.log("rendering");

  if (sessionStorage.getItem("role") === null) {
    roomId = uuidv4();
    console.log("generate id");
    // if the role is null then we will create a new room and assign the room Id to session storage
    // sessionStorage.setItem("role", "host");
    // console.log(window.location.port);
    // console.log(window.location.origin);
    // console.log(window.location.pathname.split("/").pop());
  } else {
    // get roomId and redirect back to room page
    sessionStorage.getItem("roomId");
  }

  useEffect(() => {
    setSocket(
      io.connect(window.location.origin.replace(window.location.port, "5000"))
    );
  }, []);

  // eslint-disable-next-line consistent-return
  useEffect(() => {
    if (socket) {
      console.log("on fired");
      socket.emit("board:create", roomId);
      console.log("emit fired");
      socket.on("create:complete", (msg) => {
        console.log(msg);
        window.location.assign(`/board/${roomId}`);
      });
    }
  }, [roomId, socket]);

  return null;
};

export default NetworkManager;
export { generateRoll, sendRoll, getRoll };
