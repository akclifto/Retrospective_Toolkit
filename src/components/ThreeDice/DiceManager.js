import React, { Suspense, useEffect, useMemo } from "react";
import { atom, useAtom } from "jotai";
import PropTypes from "prop-types";
import { BoxBufferGeometry } from "three";
import { uniqueImageSet } from "../Dice/Dice";
import ThemedDie from "./ThemedDie";
import CollisionMesh from "./CollisionMesh";
import { diceDefaultState } from "./gameState";

const dieOneImageState = atom([]);
const dieTwoImageState = atom([]);
const dieThreeImageState = atom([]);
const dieFourImageState = atom([]);
const dieFiveImageState = atom([]);

const rerollDieOne = atom(false);
const rerollDieTwo = atom(false);
const rerollDieThree = atom(false);
const rerollDieFour = atom(false);
const rerollDieFive = atom(false);

/* istanbul ignore next */
const DiceManager = (props) => {
  const { reroll } = props;
  const geom = useMemo(() => new BoxBufferGeometry(), []);

  const [dicePosition] = useAtom(diceDefaultState);
  const [dieImagesOne, setImagesOne] = useAtom(dieOneImageState);
  const [dieImagesTwo, setImagesTwo] = useAtom(dieTwoImageState);
  const [dieImagesThree, setImagesThree] = useAtom(dieThreeImageState);
  const [dieImagesFour, setImagesFour] = useAtom(dieFourImageState);
  const [dieImagesFive, setImagesFive] = useAtom(dieFiveImageState);

  const [rerollOne, toggleRerollOne] = useAtom(rerollDieOne);
  const [rerollTwo, toggleRerollTwo] = useAtom(rerollDieTwo);
  const [rerollThree, toggleRerollThree] = useAtom(rerollDieThree);
  const [rerollFour, toggleRerollFour] = useAtom(rerollDieFour);
  const [rerollFive, toggleRerollFive] = useAtom(rerollDieFive);

  const imageArray = [
    { images: dieImagesOne, setImages: setImagesOne },
    { images: dieImagesTwo, setImages: setImagesTwo },
    { images: dieImagesThree, setImages: setImagesThree },
    { images: dieImagesFour, setImages: setImagesFour },
    { images: dieImagesFive, setImages: setImagesFive },
  ];

  const rerollArray = [
    { rerollDie: rerollOne, reroll: toggleRerollOne },
    { rerollDie: rerollTwo, reroll: toggleRerollTwo },
    { rerollDie: rerollThree, reroll: toggleRerollThree },
    { rerollDie: rerollFour, reroll: toggleRerollFour },
    { rerollDie: rerollFive, reroll: toggleRerollFive },
  ];

  useEffect(() => {
    setImagesOne(uniqueImageSet());
    setImagesTwo(uniqueImageSet());
    setImagesThree(uniqueImageSet());
    setImagesFour(uniqueImageSet());
    setImagesFive(uniqueImageSet());
  }, [
    setImagesFive,
    setImagesFour,
    setImagesOne,
    setImagesThree,
    setImagesTwo,
  ]);

  return (
    <Suspense fallback={null}>
      {dicePosition.map((pos, index) => (
        <ThemedDie
          key={pos.uuid}
          dicePos={pos.position}
          rerollAllToggle={reroll}
          rerollValue={rerollArray[index].rerollDie}
          rerollDieToggle={rerollArray[index].reroll}
          imageSet={imageArray[index].images}
          setImages={imageArray[index].setImages}
          geom={geom}
        />
      ))}
      <CollisionMesh />
    </Suspense>
  );
};

DiceManager.propTypes = {
  reroll: PropTypes.bool.isRequired,
};

export default DiceManager;
