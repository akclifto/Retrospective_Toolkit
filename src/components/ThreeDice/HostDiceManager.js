import React, { Suspense, useEffect, useMemo, useState } from "react";
import { useAtom } from "jotai";
import PropTypes from "prop-types";
import { BoxBufferGeometry } from "three";
import { useThree, useFrame } from "@react-three/fiber";
import { uniqueImageSet } from "../Dice/Dice";
import HostThemedDie from "./HostThemedDie";
import CollisionMesh from "./CollisionMesh";
import { diceDefaultState } from "./gameState";

/* istanbul ignore next */
const newRotationArray = () => [Math.random(), Math.random(), Math.random()];

/* istanbul ignore next */
const DiceManager = (props) => {
  const {
    reroll,
    setOrbitControl,
    socket,
    roomId,
    gameStatus,
    rollSound,
  } = props;
  const geom = useMemo(() => new BoxBufferGeometry(), []);

  const [dicePosition] = useAtom(diceDefaultState);

  const [userGameReady, setUserReady] = useState(gameStatus);
  const [initialRoll, setInitialRoll] = useState(true);

  const [dieImagesOne, setImagesOne] = useState(uniqueImageSet());
  const [dieImagesTwo, setImagesTwo] = useState(uniqueImageSet());
  const [dieImagesThree, setImagesThree] = useState(uniqueImageSet());
  const [dieImagesFour, setImagesFour] = useState(uniqueImageSet());
  const [dieImagesFive, setImagesFive] = useState(uniqueImageSet());

  const [rerollOne, toggleRerollOne] = useState(false);
  const [rerollTwo, toggleRerollTwo] = useState(false);
  const [rerollThree, toggleRerollThree] = useState(false);
  const [rerollFour, toggleRerollFour] = useState(false);
  const [rerollFive, toggleRerollFive] = useState(false);

  const [rotationOne, setRotationOne] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationTwo, setRotationTwo] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationThree, setRotationThree] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationFour, setRotationFour] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationFive, setRotationFive] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);

  const imageArray = [
    { images: dieImagesOne, setImages: setImagesOne },
    { images: dieImagesTwo, setImages: setImagesTwo },
    { images: dieImagesThree, setImages: setImagesThree },
    { images: dieImagesFour, setImages: setImagesFour },
    { images: dieImagesFive, setImages: setImagesFive },
  ];

  const userImageArray = [
    dieImagesOne,
    dieImagesTwo,
    dieImagesThree,
    dieImagesFour,
    dieImagesFive,
  ];

  const userRotationArray = [
    rotationOne,
    rotationTwo,
    rotationThree,
    rotationFour,
    rotationFive,
  ];

  const rerollArray = [
    { rerollDie: rerollOne, reroll: toggleRerollOne },
    { rerollDie: rerollTwo, reroll: toggleRerollTwo },
    { rerollDie: rerollThree, reroll: toggleRerollThree },
    { rerollDie: rerollFour, reroll: toggleRerollFour },
    { rerollDie: rerollFive, reroll: toggleRerollFive },
  ];

  const rotationArray = [
    { rotation: rotationOne, setRotation: setRotationOne },
    { rotation: rotationTwo, setRotation: setRotationTwo },
    { rotation: rotationThree, setRotation: setRotationThree },
    { rotation: rotationFour, setRotation: setRotationFour },
    { rotation: rotationFive, setRotation: setRotationFive },
  ];

  useEffect(() => {
    if (!initialRoll && userGameReady) {
      const newImageArray = [];
      const newRotArray = [];

      for (let i = 0; i < 5; i += 1) {
        newImageArray.push(uniqueImageSet());
        imageArray[i].setImages(newImageArray[i]);
        newRotArray.push(newRotationArray());
        rotationArray[i].setRotation(newRotArray[i]);
      }

      socket.emit("host:newRoll", roomId, newRotArray, newImageArray);
    } else if (!userGameReady) {
      socket.emit("game:start", roomId, userRotationArray, userImageArray);
      setUserReady(true);
      setInitialRoll(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [reroll]);

  useEffect(() => {
    if (userGameReady) {
      socket.emit("get:update", roomId);
      setInitialRoll(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    socket.on("user:init", (rotationValues, imagesArray) => {
      // eslint-disable-next-line array-callback-return
      imagesArray.map((image, index) => {
        imageArray[index].setImages(image);
        rotationArray[index].setRotation(rotationValues[index]);
      });
    });
    socket.on("user:initQueue", (rollObject) => {
      if (rollObject.die === null) {
        // eslint-disable-next-line array-callback-return
        rollObject.image.map((image, index) => {
          imageArray[index].setImages(image);
          rotationArray[index].setRotation(rollObject.rotation[index]);
        });
      } else {
        // eslint-disable-next-line array-callback-return
        imageArray[rollObject.die].setImages(rollObject.image);
        rotationArray[rollObject.die].setRotation(rollObject.rotation);
        rollSound();
      }
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (!initialRoll) {
      const newImage = uniqueImageSet();
      const newRotation = newRotationArray();
      imageArray[0].setImages(newImage);
      rotationArray[0].setRotation(newRotation);
      socket.emit("host:newDieRoll", roomId, newRotation, newImage, 0);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rerollOne]);

  useEffect(() => {
    if (!initialRoll) {
      const newImage = uniqueImageSet();
      const newRotation = newRotationArray();
      imageArray[1].setImages(newImage);
      rotationArray[1].setRotation(newRotation);
      socket.emit("host:newDieRoll", roomId, newRotation, newImage, 1);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rerollTwo]);

  useEffect(() => {
    if (!initialRoll) {
      const newImage = uniqueImageSet();
      const newRotation = newRotationArray();
      imageArray[2].setImages(newImage);
      rotationArray[2].setRotation(newRotation);
      socket.emit("host:newDieRoll", roomId, newRotation, newImage, 2);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rerollThree]);

  useEffect(() => {
    if (!initialRoll) {
      const newImage = uniqueImageSet();
      const newRotation = newRotationArray();
      imageArray[3].setImages(newImage);
      rotationArray[3].setRotation(newRotation);
      socket.emit("host:newDieRoll", roomId, newRotation, newImage, 3);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rerollFour]);

  useEffect(() => {
    if (!initialRoll) {
      const newImage = uniqueImageSet();
      const newRotation = newRotationArray();
      imageArray[4].setImages(newImage);
      rotationArray[4].setRotation(newRotation);
      socket.emit("host:newDieRoll", roomId, newRotation, newImage, 4);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rerollFive]);

  const { viewport } = useThree();
  const mousePos = [];

  useFrame((state) => {
    const { mouse } = state;
    const { width } = viewport;
    const { height } = viewport;
    mousePos[0] = (mouse.x * width) / 2;
    mousePos[1] = (mouse.y * height) / 2;
  });

  return (
    <Suspense fallback={null}>
      {dicePosition.map((pos, index) => (
        <HostThemedDie
          key={pos.uuid}
          diceInitPos={pos.position}
          rerollValue={rerollArray[index].rerollDie}
          rerollDieToggle={rerollArray[index].reroll}
          imageSet={imageArray[index].images}
          geom={geom}
          setOrbitControl={setOrbitControl}
          mousePos={mousePos}
          rotationValues={rotationArray[index].rotation}
        />
      ))}
      <CollisionMesh />
    </Suspense>
  );
};

DiceManager.propTypes = {
  reroll: PropTypes.bool,
  setOrbitControl: PropTypes.func.isRequired,
  roomId: PropTypes.string.isRequired,
  gameStatus: PropTypes.bool.isRequired,
  rollSound: PropTypes.func.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  socket: PropTypes.object.isRequired,
};

DiceManager.defaultProps = {
  reroll: false,
};

export default DiceManager;
