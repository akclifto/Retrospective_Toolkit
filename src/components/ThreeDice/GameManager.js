import React, { Suspense } from "react";
import { Html, useGLTF, useProgress } from "@react-three/drei";
import Button from "@material-ui/core/Button";
import { makeStyles } from "@material-ui/core/styles";
import Icon from "@material-ui/core/Icon";
import { useAtom } from "jotai";
import PropTypes from "prop-types";
import { gameStartState, rerollState } from "./gameState";
import DiceManager from "./DiceManager";

/* istanbul ignore next */
const GameManager = (props) => {
  const { setOrbitControl } = props;
  const [gameStarted, setGameState] = useAtom(gameStartState);
  const [reroll, rerollDice] = useAtom(rerollState);
  const rollSound = new Audio("/diceRoll.m4a");
  const useStyles = makeStyles((theme) => ({
    button: {
      margin: theme.spacing(1),
      width: "120px",
    },
    buttonStart: {
      margin: theme.spacing(1),
      width: "160px",
    },
  }));

  const classes = useStyles();

  return (
    <>
      <ambientLight intensity={0.5} />
      <directionalLight
        intensity={1.0}
        position={[0, 20, 0]}
        shadow-mapSize-width={1024}
        shadow-mapSize-height={1024}
      />
      <Suspense fallback={<ProgressBar />}>
        <ModelLoader url="table/BlackRedTable.glb" />
      </Suspense>
      {!gameStarted && (
        <Html position={[-4, 0, 2]} scaleFactor={25}>
          <Button
            variant="contained"
            color="primary"
            className={classes.buttonStart}
            endIcon={<Icon>casino</Icon>}
            onClick={() => {
              setGameState(true);
              rollSound.play();
            }}
          >
            Start Game
          </Button>
        </Html>
      )}
      {gameStarted && (
        <>
          <DiceManager reroll={reroll} setOrbitControl={setOrbitControl} />
          <Html position={[-3, 0, 7]} scaleFactor={25}>
            <Button
              variant="contained"
              color="primary"
              className={classes.button}
              endIcon={<Icon>casino</Icon>}
              onClick={() => {
                rerollDice(!reroll);
                rollSound.play();
              }}
            >
              Roll It!
            </Button>
          </Html>
        </>
      )}
    </>
  );
};
GameManager.propTypes = {
  setOrbitControl: PropTypes.func.isRequired,
};
/* istanbul ignore next */
const ProgressBar = () => {
  const { progress } = useProgress();
  return <Html center>{Math.trunc(progress)} % loaded</Html>;
};
/* istanbul ignore next */
const ModelLoader = ({ url }) => {
  const { scene } = useGLTF(url);
  return <primitive object={scene} dispose={null} />;
};
ModelLoader.propTypes = {
  url: PropTypes.string.isRequired,
};

export default GameManager;
