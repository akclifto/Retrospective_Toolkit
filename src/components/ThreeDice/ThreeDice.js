import React, { useEffect } from "react";
import { Canvas } from "react-three-fiber";
import { OrbitControls } from "@react-three/drei";
import { Physics } from "@react-three/cannon";
import { Provider } from "jotai";
import { initDiceImages } from "../Dice/Dice";
import GameManager from "./GameManager";

const ThreeDice = () => {
  // Allows the initDiceImages function to load only once on startup.
  useEffect(() => {
    const loadDice = async () => {
      await initDiceImages();
    };
    loadDice();
  }, []);
  return (
    <Canvas
      concurrent
      style={{ width: "100vw", height: "70vh" }}
      camera={{ position: [0, 20, 8], fov: 50 }}
    >
      <Provider>
        <Physics gravity={[0, -30, 0]} defaultContactMaterial>
          <GameManager />
        </Physics>
      </Provider>
      <OrbitControls />
    </Canvas>
  );
};

export default ThreeDice;
