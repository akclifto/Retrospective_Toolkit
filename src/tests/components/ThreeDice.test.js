import React from "react";
import { shallow } from "enzyme";
import ThreeDice from "../../components/ThreeDice/ThreeDice";

let consoleSpy;

beforeEach(() => {
  // catch jest console error and dump into empty function.  Catches react-three physics console error.
  consoleSpy = jest.spyOn(console, "error").mockImplementation(() => {});
});

describe("ThreeDice Testing", () => {
  it("renders and matches snapshot", () => {
    expect(shallow(<ThreeDice />)).toMatchSnapshot();
    expect(consoleSpy).toHaveBeenCalled();
    expect(consoleSpy).toHaveBeenCalledTimes(1);
  });
});
