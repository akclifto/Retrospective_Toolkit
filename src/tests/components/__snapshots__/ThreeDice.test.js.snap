// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeDice Testing renders and matches snapshot 1`] = `
<Fragment>
  <Memo(Canvas)
    camera={
      Object {
        "fov": 50,
        "position": Array [
          0,
          20,
          8,
        ],
      }
    }
    concurrent={true}
    invalidateFrameloop={true}
    style={
      Object {
        "height": "70vh",
        "width": "100vw",
      }
    }
  >
    <ForwardRef
      center={true}
    >
      Loading game textures....
    </ForwardRef>
  </Memo(Canvas)>
  )
</Fragment>
`;
